src/login/logind-dbus.c

```diff
@@ -676,36 +676,60 @@ static int method_list_inhibitors(sd_bus_message *message, void *userdata, sd_bu
         return sd_bus_send(NULL, reply, NULL);
 }
 
-static int method_create_session(sd_bus_message *message, void *userdata, sd_bus_error *error) {
-        const char *service, *type, *class, *cseat, *tty, *display, *remote_user, *remote_host, *desktop;
+static int create_session(sd_bus_message *message,
+                          void *userdata,
+                          sd_bus_error *error,
+                          uid_t uid,
+                          pid_t pid,
+                          int pidfd,
+                          const char *service,
+                          const char *type,
+                          const char *class,
+                          const char *desktop,
+                          const char *cseat,
+                          uint32_t vtnr,
+                          const char *tty,
+                          const char *display,
+                          int remote,
+                          const char *remote_user,
+                          const char *remote_host
+                          ) {
         _cleanup_free_ char *id = NULL;
         Session *session = NULL;
         uint32_t audit_id = 0;
         Manager *m = ASSERT_PTR(userdata);
         User *user = NULL;
         Seat *seat = NULL;
-        pid_t leader;
-        uid_t uid;
-        int remote;
-        uint32_t vtnr = 0;
+        pid_t leader = 0;
         SessionType t;
         SessionClass c;
         int r;
 
         assert(message);
 
-        assert_cc(sizeof(pid_t) == sizeof(uint32_t));
-        assert_cc(sizeof(uid_t) == sizeof(uint32_t));
-
-        r = sd_bus_message_read(message, "uusssssussbss",
-                                &uid, &leader, &service, &type, &class, &desktop, &cseat,
-                                &vtnr, &tty, &display, &remote, &remote_user, &remote_host);
-        if (r < 0)
-                return r;
-
         if (!uid_is_valid(uid))
                 return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid UID");
-        if (leader < 0 || leader == 1 || leader == getpid_cached())
+
+        if (pidfd > 0) {
+                r = pidfd_get_pid(pidfd, &leader);
+                if (r < 0)
+                        return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "Failed to get PID from PID fd");
+        } else if (pid == 0) {
+                _cleanup_(sd_bus_creds_unrefp) sd_bus_creds *creds = NULL;
+
+                r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID, &creds);
+                if (r < 0)
+                        return r;
+
+                r = sd_bus_creds_get_pid(creds, (pid_t*) &leader);
+                if (r < 0)
+                        return r;
+        } else if (pid > 0)
```

Comment:
this will always hold. hence change this to:

```c
} else {
        assert(pid > 0);
        leader = pid;
}
```

this encodes expectations, and acts as a comment of sorts, too